### ISSUE TITLE
Please bump wgpu to `0.16.1` to fix unsoundness on wasm32/web-sys targets

### ISSUE THREAD

**Issue by 197g** [BVE-Reborn/rend3#487]
Specifically referring to any release with this commit:

<https://github.com/gfx-rs/wgpu/commit/cfc038aeb9b732019dc03fa7eeaac1c7f37fc71f>

For instance this bug might materialize during device enumeration. Since `get_info` is called, where we will enter

```rust
    fn adapter_get_info(&self, adapter: &ObjectId, adapter_data: &crate::Data) -> AdapterInfo {
        let adapter = <T::AdapterId>::from(*adapter);
        let adapter_data = downcast_ref(adapter_data);
        Context::adapter_get_info(self, &adapter, adapter_data)
    }
```

The unsoundness is in the call to `<T::AdapterId>::from` which ends up incorrectly _taking ownership_ of the adapter which was passed only as a mutable reference. The `JsValue` representing the adapter is then dropped on function exit, but also later used further in `rend3/src/setup.rs:create_iad`. Dropping the value has the effect of invoking the corresponding web-sys binding such as `__wbindgen_object_drop_ref`. The later use is then use-after-free (and does quite frankly funny things such as trying to access the `limits` property from the integer that wasm-bindgen uses to mark the free heap slot).

That was quite fun to debug, learned a lot of the wasm32 ABI in the process :joy:

<details>
<summary>Full dissassembly of the improper adapter_get_info</summary>

```
(; find out why we call __wbindgen_object_drop_ref below.
 ; this causes the adaptor to be dropped but we only ever had a reference to it?
 ; which results in the following obviously bad sequence in wasm-bindgen:
 ;
 ; _ = take(134);
 ; heap[192] = __wbg_limits_254f53e662b2a6f2(134); // [= undefined]
 ; __wbg_maxTextureDimension1D_7ad88ba70060cbc0(192); // fails
 ;
 ; This sequence is the proper from [rend3/src/setup.rs:create_iad]
 ;          let info = adapter.get_info();
 ;          let limits = adapter.limits();
 ;
 ; Here's the supposedly (?) bad function

  ; Probably from this source (wgpu@.15.1:source/src/context.rs)
  ;
  ; Note: crate::Data is dyn Any + Send + Sync
  ; Note: $var2 is the ObjectId with its br_if $label0 the unwrap in `fn ObjectId::id(&self) -> NonZeroU32`.
  ;
  ; fn adapter_get_info(&self, adapter: &ObjectId, adapter_data: &crate::Data) -> AdapterInfo {
  ;     let adapter = <T::AdapterId>::from(*adapter);
  ;     let adapter_data = downcast_ref(adapter_data);
  ;     Context::adapter_get_info(self, &adapter, adapter_data)
  ; }
  (func $<T as wgpu::context::DynContext>::adapter_get_info::h748a0159be521005 (;3147;) (param $var0 i32) (param $var1 i32) (param $var2 i32) (param $var3 i32) (param $var4 i32)
    (local $var5 i64)
    block $label0
      local.get $var2
      i64.load
      local.tee $var5
      i64.const 0
      i64.eq
      br_if $label0
      local.get $var0
      i64.const 4294967296
      ; Thought process: $var0 is the return parameter, I think?
      i64.store offset=8 align=4
      local.get $var0
      i64.const 0
      i64.store align=4
      local.get $var0
      i32.const 6
      i32.store16 offset=44
      local.get $var0
      i64.const 4294967296
      i64.store offset=20 align=4
      local.get $var0
      i32.const 16
      i32.add
      i32.const 0
      i32.store
      local.get $var0
      i32.const 36
      i32.add
      i64.const 1
      i64.store align=4
      local.get $var0
      i32.const 28
      i32.add
      i64.const 0
      i64.store align=4
      block $label1
        ; from local.tee $var5 which was i64.load (local.get $var2)
        local.get $var5
        i32.wrap_i64
        local.tee $var0
        i32.const 132
        i32.lt_u
        br_if $label1
        local.get $var0
        call $wasm_bindgen::__wbindgen_object_drop_ref::h6e014cdb98ab112f
      end $label1
      return
    end $label0
    i32.const 1560034
    i32.const 43
    i32.const 1560276
    call $core::panicking::panic::h72f0442034a99972
    unreachable
  )
)
```

</details>

**Comment by 197g** [comment]
Grain of salt, wasm32-wasi is not supported wasm-bindgen so I'm using the work-in-progress patch for it. Maybe something in there is also causing this, not wgpu implementation. It's kind of late.

**Comment by cwfitzgerald** [comment]
Closed by #491
